src/lsm/zig_zag_merge.zig

```diff
@@ -0,0 +1,664 @@
+pub fn ZigZagMergeIteratorType(
+    comptime Context: type,
+    comptime Key: type,
+    comptime Value: type,
+    comptime key_from_value: fn (*const Value) callconv(.Inline) Key,
+    comptime streams_max: u32,
+    /// Peek the next key in the stream identified by `stream_index`.
+    /// For example, `peek(stream_index=2)` returns `user_streams[2][0]`.
+    /// Returns `Drained` if the stream was consumed and must be refilled
+    /// before calling `peek()` again.
+    /// Returns `Empty` if the stream was fully consumed and reached the end.
+    comptime stream_peek: fn (
+        context: *Context,
+        stream_index: u32,
+    ) error{ Empty, Drained }!Key,
+    /// Consumes the current value and moves the stream identified by `stream_index`.
+    /// Pop is always called after `peek()`, it is not expected that the stream be `Empty`
+    /// or `Drained`.
+    comptime stream_pop: fn (context: *Context, stream_index: u32) Value,
+    /// Probes the stream identified by `stream_index` causing it to move to the next value such
+    /// as `>=key` or `<=key` depending on the iterator direction.
+    /// Should not be called when the current key already matches the probe.
+    /// The stream may become `Empty` or `Drained` _after_ probing.
+    comptime stream_probe: fn (context: *Context, stream_index: u32, key: Key) void,
+) type {
+    return struct {
+        const ZigZagMergeIterator = @This();
+        const BitSet = std.bit_set.IntegerBitSet(streams_max);
+
+        context: *Context,
+        streams_count: u32,
+        direction: Direction,
+        previous_key_popped: ?Key = null,
+
+        /// At least two scans are required for zig-zag merge.
+        pub fn init(
+            context: *Context,
+            streams_count: u32,
+            direction: Direction,
+        ) ZigZagMergeIterator {
+            assert(streams_count <= streams_max);
+            assert(streams_count > 1);
+
+            return .{
+                .context = context,
+                .streams_count = streams_count,
+                .direction = direction,
+            };
+        }
+
+        // Resets the iterator when the underlying streams are moved.
+        // It's not necessary for ZigZagMerge, but it follows the same API for all MergeIterators.
+        pub fn reset(it: *ZigZagMergeIterator) void {
+            _ = it;
+        }
+
+        pub fn pop(it: *ZigZagMergeIterator) error{Drained}!?Value {
+            while (try it.peek_key()) |key| {
+                const value = stream_pop(it.context, 0);
+                assert(key_from_value(&value) == key);
+                for (1..it.streams_count) |stream_index| {
+                    const value_other = stream_pop(it.context, @intCast(stream_index));
+                    assert(key_from_value(&value_other) == key);
+
+                    if (constants.verify) {
+                        // It's assumed that streams will produce the same value.
```

Comment:
> Will we later use zig-zag merge iterator with Key != Value?

Yes, scans produce only timestamps, and since the composite key prefix is always fixed (equality), they also take only the timestamp as the key, so `Key == Value` for merging scans.

I don't think we should enforce `Key == Value`, the intersection logic can be reused as a primitive.